src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -489,18 +494,18 @@ internal TypeWithAnnotations SubstituteTypeCore(AbstractTypeMap typeMap)
             {
                 newAnnotation = NullableAnnotation;
             }
-            else if (NullableAnnotation != NullableAnnotation.Oblivious)
+            else if (NullableAnnotation.IsNotAnnotated())
             {
-                if (!typeSymbol.IsTypeParameterDisallowingAnnotationInCSharp8())
+                if (typeSymbol.IsTypeParameterDisallowingAnnotationInCSharp8() && !newTypeWithModifiers.Type.IsTypeParameterDisallowingAnnotationInCSharp8())
                 {
-                    newAnnotation = NullableAnnotation;
+                    newAnnotation = newTypeWithModifiers.NullableAnnotation;
                 }
                 else
                 {
-                    newAnnotation = newTypeWithModifiers.NullableAnnotation;
+                    newAnnotation = NullableAnnotation;
```

Comment:
>newAnnotation = NullableAnnotation; [](start = 20, length = 35)

If new annotation is oblivious, shouldn't that always win for ```typeSymbol.IsTypeParameterDisallowingAnnotationInCSharp8()```?

Reply:
Trying to answer this question, I tried to spec out the behavior of this method to make sure I understand it.

I wrote the matrix that I think we want: [https://gist.github.com/jcouv/4dd5291e65a4cbffa83c03440427d557](https://gist.github.com/jcouv/4dd5291e65a4cbffa83c03440427d557)
Took a look with Chuck and made some small adjustments.
I would summarize this matrix as: (1) annotated original or substitution type yields annotated result, (2) otherwise, use the annotation from substitution type.

@AlekseyTs, does that look right to you too? I've included a legend to help decode that matrix, but would be happy to hop on a call to walk you through.

---
In reply to: [503339925](https://github.com/dotnet/roslyn/pull/48270#discussion_r503339925) [](ancestors = 503339925)